core/tracker/incldelay.go

```diff
@@ -6,112 +6,211 @@ import (
 	"context"
 	"fmt"
 	"sync"
+	"time"
+
+	eth2p0 "github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/jonboulle/clockwork"
 
 	"github.com/obolnetwork/charon/app/errors"
 	"github.com/obolnetwork/charon/app/eth2wrap"
+	"github.com/obolnetwork/charon/app/log"
+	"github.com/obolnetwork/charon/app/z"
 	"github.com/obolnetwork/charon/core"
 )
 
-// inclDelayLag is the number of slots to lag before calculating inclusion delay.
-// Half an epoch is good compromise between finality and small gaps on startup.
-const inclDelayLag = 16
+const (
+	// inclDelayLag is the number of slots to lag before calculating inclusion delay.
+	// Half an epoch is good compromise between finality and small gaps on startup.
+	inclDelayLag = 16
+	// trimEpochOffset is the number of epochs after which we delete cached broadcast delays.
+	// This matches scheduler trimEpochOffset.
+	trimEpochOffset = 3
+)
 
 // dutiesFunc returns the duty definitions for a given duty.
 type dutiesFunc func(context.Context, core.Duty) (core.DutyDefinitionSet, error)
 
-// NewInclDelayFunc returns a function that calculates attestation inclusion delay for a block.
+// NewInclusionDelay returns a new inclusion delay tracker.
+func NewInclusionDelay(ctx context.Context, eth2Cl eth2wrap.Client, dutiesFunc dutiesFunc) (*InclusionDelay, error) {
+	genesisTime, err := eth2Cl.GenesisTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	slotDuration, err := eth2Cl.SlotDuration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	slotsPerEpoch, err := eth2Cl.SlotsPerEpoch(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &InclusionDelay{
+		eth2Cl:         eth2Cl,
+		dutiesFunc:     dutiesFunc,
+		instrumentFunc: instrumentAvgDelay,
+		logMappingFunc: logMapping,
+		bcastDelays:    make(map[int64]map[eth2p0.Root]time.Duration),
+		genesisTime:    genesisTime,
+		slotDuration:   slotDuration,
+		slotsPerEpoch:  int(slotsPerEpoch),
+		clock:          clockwork.NewRealClock(),
+	}, nil
+}
+
+// InclusionDelay tracks the inclusion delay of attestations.
+type InclusionDelay struct {
+	eth2Cl         eth2wrap.Client
+	dutiesFunc     dutiesFunc
+	instrumentFunc func(inclDelaySlots []int64)
+	logMappingFunc func(ctx context.Context, slot int64, bcastDelay time.Duration, inclDelay int64, isAggregated bool)
+	genesisTime    time.Time
+	slotDuration   time.Duration
+	slotsPerEpoch  int
+	clock          clockwork.Clock
+
+	mu            sync.Mutex
+	bcastDelays   map[int64]map[eth2p0.Root]time.Duration
```

Comment:
so we are storing broadcast delays for attestations per slot per block root, but we can submit multiple attestations with the same slot and block root. 

This map will only take care of one of those attestation. I think if we are tracking then we should track all of them. Maybe modify the map as `map[int64]map[eth2p0.BLSPubKey]time.Duration` where we can get the pubkey from duty.

Reply:
Also while iterating all the attestation in the block this broadcast might not EXACTLY map with the desired attestation if we have submitted multiple attestations with the same and block root.